lib/msal-common/src/client/BaseClient.ts

```diff
@@ -79,85 +75,60 @@ export abstract class BaseClient {
 
-    /**
-     * Creates a url for logging in a user. This will by default append the client id to the list of scopes,
-     * allowing you to retrieve an id token in the subsequent code exchange. Also performs validation of the request parameters.
-     * Including any SSO parameters (account, sid, login_hint) will short circuit the authentication and allow you to retrieve a code without interaction.
-     * @param request
-     */
-    abstract async createLoginUrl(request: AuthenticationParameters): Promise<string>;
+        // Default authority instance.
+        this.defaultAuthorityInstance = AuthorityFactory.createInstance(
+            this.config.authOptions.authority || Constants.DEFAULT_AUTHORITY,
+            this.networkClient
+        );
+    }
 
     /**
-     * Creates a url for logging in a user. Also performs validation of the request parameters.
-     * Including any SSO parameters (account, sid, login_hint) will short circuit the authentication and allow you to retrieve a code without interaction.
-     * @param request
+     * Create authority instance if not set already, resolve well-known-endpoint
+     * @param authorityString
      */
-    abstract async createAcquireTokenUrl(request: AuthenticationParameters): Promise<string>;
+    protected async createAuthority(authorityString: string): Promise<Authority> {
 
-    /**
-     * Handles the hash fragment response from public client code request. Returns a code response used by
-     * the client to exchange for a token in acquireToken.
-     * @param hashFragment
-     */
-    abstract handleFragmentResponse(hashFragment: string): CodeResponse;
+        // TODO expensive to resolve authority endpoints everytime.
+        const authority: Authority = authorityString
+            ? AuthorityFactory.createInstance(authorityString, this.networkClient)
+            : this.defaultAuthorityInstance;
 
-    /**
-     * Given an authorization code, it will perform a token exchange using cached values from a previous call to
-     * createLoginUrl() or createAcquireTokenUrl(). You must call this AFTER using one of those APIs first. You should
-     * also use the handleFragmentResponse() API to pass the codeResponse to this function afterwards.
-     * @param codeResponse
-     */
-    abstract async acquireToken(codeResponse: CodeResponse): Promise<TokenResponse>;
+        try {
+            await authority.resolveEndpointsAsync();
+        } catch (error) {
+            throw ClientAuthError.createEndpointDiscoveryIncompleteError(error);
+        }
 
-    /**
-     * Retrieves a token from cache if it is still valid, or uses the cached refresh token to renew
-     * the given token and returns the renewed token. Will throw an error if login is not completed (unless
-     * id tokens are not being renewed).
-     * @param request
-     */
-    abstract async renewToken(request: TokenRenewParameters): Promise<TokenResponse>;
+        return authority;
```

Comment:
authority is not a promise right ? you want to return authority.resolveEndpointsAsync

Reply:
It's an async function, so authority get's wrapped in a promise.